compiler/rustc_data_structures/src/statistics.rs

```diff
@@ -0,0 +1,164 @@
+//! # Support for collecting simple statistics
+//!
+//! Statistics are useful for collecting metrics from optimization passes, like
+//! the number of simplifications performed. To avoid introducing overhead, the
+//! collection of statistics is enabled only when rustc is compiled with
+//! debug-assertions.
+//!
+//! Statistics are static variables defined in the module they are used, and
+//! lazy registered in the global collector on the first use. Once registered,
+//! the collector will obtain their values at the end of compilation process
+//! when requested with -Zmir-opt-stats option.
+
+use parking_lot::{const_mutex, Mutex};
+use std::fmt::Write as _;
+use std::io::{self, Write as _};
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+
+static COLLECTOR: Collector = Collector::new();
```

Comment:
This should be part of `Session` or `TyCtxt`. There may be multiple rustc sessions running in a single process. (for example RLS)

Reply:
I experimented with different implementation approaches:

1. Statistics definitions are in the context together with their values.
2. Statistics definitions are in MIR passes, their values are stored in the context.
3. Statistics definitions and values are stored together with MIR passes.

I strongly prefer variants that put statistics definitions together with MIR passes that uses them. In that approach, adding and removing counters is trivial, additionally any unused counters are detected as such during compilation.

Values can be stored in the context if we want to support multiple rustc sessions in a single process, although is requires passing context to all those places where counters are used. But is there a any real use-case? Why would you use rustc with debug-assertions and MIR optimizations counters inside the RLS?

Reply:
I agree with @bjorn3. We've tried pretty hard to move these sorts of things into `Session` and `TyCtxt` to allow running multiple sessions in the same process (both for RLS and eventually for one rustc process compiling multiple crates simultaneously). 

Since this is essentially unstable, diagnostic data and doesn't effect the compilation artifacts, it may be worth bending that rule because, as you point out, having the counters defined in their MIR passes is quite nice. I don't personally feel comfortable approving that since this is a cross-cutting concern for the compiler as a whole and there may be others on the compiler team that have strong opinions about breaking that rule.